### PR TITLE
Use `mslex.quote` when shell-escaping strings

### LINKED ISSUE #6585: [BUG] unable to pass artifacts between steps in multistep workflow
### Willingness to contribute

No. I cannot contribute a bug fix at this time.

### MLflow version

1.27.0

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10
- **Python version**: 3.10.4
- **yarn version, if running the dev UI**:

### Describe the problem

I'm trying to run a multistep mlflow project.  When trying to use artifacts generated in an earlier step and stored in an S3 bucket in a subsequent step, the files cannot be loaded due to the fact that the file path argument to the artifacts are passed in nested quotes.  It causes the following error: 
OSError: [Errno 22] Invalid argument.  I would expect a normal file path to be passed to the downstream step so the file can be read in and used.  

### Tracking information

_No response_

### Code to reproduce issue

```python
import mlflow
import argparse

parser = argparse.ArgumentParser(argument_default=argparse.SUPPRESS)
parser.add_argument("--fully_preprocessed_train_data", type=str, default="fully_preprocessed_train_data.pkl")
parser.add_argument("--fully_preprocessed_test_data", type=str, default="fully_preprocessed_test_data.pkl")
parser.add_argument("--fully_preprocessed_oot_data", type=str, default="fully_preprocessed_oot_data.pkl")
args = parser.parse_args()

preprocessing_params = {raw_data: raw_data_filepath}
train_params = {}

def workflow():
    with mlflow.start_run(description="Running workflow") as run:
        print("Performing custom preprocessing")
        custom_preprocessing_run = mlflow.run(".", "custom_preprocess", parameters=preprocessing_params)
        custom_preprocessing_runid = mlflow.tracking.MlflowClient().get_run(custom_preprocessing_run.run_id)
        data_path_uri = custom_preprocessing_runid.info.artifact_uri
        print(f"Sourcing training data from {data_path_uri}")
        mlflow.set_tag(key="custom_preprocess_runid", value=custom_preprocessing_runid)
        train_params["preprocessed_train_data_path"] = os.path.join(data_path_uri, args.fully_preprocessed_train_data).replace("\\", 
        "/")
        train_params["preprocessed_test_data_path"] = os.path.join(data_path_uri, args.fully_preprocessed_test_data).replace("\\", 
        "/")
        train_params["preprocessed_oot_data_path"] = os.path.join(data_path_uri, args.fully_preprocessed_oot_data).replace("\\", 
        "/")
        print("Launching training script")
        train_run = mlflow.run(".", "train", parameters=train_params)

if __name__ == "__main__":
    workflow()
```

### Stack trace

```
Launching training script
2022/08/23 11:58:40 INFO mlflow.utils.conda: Conda environment mlflow-f9436806216344de765e4a9ad154d6365504135b already exists.
2022/08/23 11:58:40 INFO mlflow.projects.utils: === Created directory C:\Users\NICOLE~1\AppData\Local\Temp\tmp23zpl4w1 for downloading remote URIs passed to arguments of type 'path' ===
2022/08/23 11:59:05 INFO mlflow.projects.backend.local: === Running command 'conda activate mlflow-f9436806216344de765e4a9ad154d6365504135b && python train.py  --preprocessed_train_data_path 'C:\Users\NICOLE~1.FAR\AppData\Local\Temp\tmp23zpl4w1\param_1\fully_preprocessed_train_data.pkl' --preprocessed_test_data_path 'C:\Users\NICOLE~1\AppData\Local\Temp\tmp23zpl4w1\param_2\fully_preprocessed_test_data.pkl' --preprocessed_oot_data_path 'C:\Users\NICOLE~1\AppData\Local\Temp\tmp23zpl4w1\param_3\fully_preprocessed_oot_data.pkl'  in run with ID '8931f30fbffe411696b49dd67d77596d' ===
Traceback (most recent call last):
  File "C:\Users\nicole\Desktop\repos\data-science-projects\multistep_workflow\train.py", line 252, in <module>
    train()
  File "C:\Users\nicole\Desktop\repos\data-science-projects\multistep_workflow\train.py", line 80, in train
    train_trans = pd.read_pickle(args.preprocessed_train_data_path)
  File "C:\Users\nicole\Miniconda3\envs\mlflow-f9436806216344de765e4a9ad154d6365504135b\lib\site-packages\pandas\io\pickle.py", line 187, in read_pickle
    with get_handle(
  File "C:\Users\nicole\Miniconda3\envs\mlflow-f9436806216344de765e4a9ad154d6365504135b\lib\site-packages\pandas\io\common.py", line 798, in get_handle
    handle = open(handle, ioargs.mode)
OSError: [Errno 22] Invalid argument: "'C:\\Users\\NICOLE~1\\AppData\\Local\\Temp\\tmp23zpl4w1\\param_1\\fully_preprocessed_train_data.pkl'"
Traceback (most recent call last):
  File "C:\Users\nicole\Desktop\repos\data-science-projects\pti_multistep_workflow\main.py", line 98, in <module>
    workflow()
  File "C:\Users\nicole\Desktop\repos\data-science-projects\pti_multistep_workflow\main.py", line 93, in workflow
    train_run = mlflow.run(".", "train", parameters=train_params)
  File "C:\Users\nicole\Miniconda3\envs\mlflow-f9436806216344de765e4a9ad154d6365504135b\lib\site-packages\mlflow\projects\__init__.py", line 346, in run
    _wait_for(submitted_run_obj)
  File "C:\Users\nicole\Miniconda3\envs\mlflow-f9436806216344de765e4a9ad154d6365504135b\lib\site-packages\mlflow\projects\__init__.py", line 363, in _wait_for       
    raise ExecutionException("Run (ID '%s') failed" % run_id)
mlflow.exceptions.ExecutionException: Run (ID '8931f30fbffe411696b49dd67d77596d') failed
2022/08/23 11:59:32 ERROR mlflow.cli: === Run (ID '12d3ed0d7c3d47eb9a200c9f2ceff20c') 
failed ===
```

### Other info / logs

_No response_

### What component(s) does this bug affect?

- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [X] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [X] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <hkawamura0130@gmail.com>

## Related Issues/PRs

Fix #6585

## What changes are proposed in this pull request?

Use `mslex.quote` (https://github.com/smoofra/mslex) when shell-escaping strings.

## How is this patch tested?

- [ ] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [x] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by BenWilson2 — APPROVED** [review]
LGTM!

**Review by WeichenXu123 — COMMENTED** [review]
I think we should add unit test for `mslex_quote`. :)

File: mlflow/utils/string_utils.py [hunk 1]
```diff
@@ -61,3 +64,68 @@ def _create_table(
...
+                    yield onlyslashes
+            else:
+                yield text
+        yield '"'
+
+    return "".join(parts())
+
+
+def quote(s):
+    return mslex_quote(s) if os.name == "nt" else shlex.quote(s)
```
Comment (dbczumar) [inline mlflow/utils/string_utils.py:131]: Looks like we also use `shlex.quote()` elsewhere in `databricks.py`: https://github.com/mlflow/mlflow/blob/f65429478255b88a211d7dd262c159b4939b44b2/mlflow/projects/databricks.py#L12. Should we change this to `quote`?

**Review by dbczumar — APPROVED** [review]
LGTM with one comment - https://github.com/mlflow/mlflow/pull/6745/files#r969068670. Thanks @harupy !

File: mlflow/utils/string_utils.py (hunk 1 above)
Comment (harupy) [reply mlflow/utils/string_utils.py:131]: Thanks for the catch! Fixed.

File: tests/utils/test_string_utils.py [hunk 2]
```diff
@@ -30,3 +30,9 @@ def test_is_string_type():
     assert not is_string_type({"test": "string"})
     assert not is_string_type(12)
     assert not is_string_type(12.7)
+
+
+def test_mslex_quote():
+    assert mslex_quote("abc") == "abc"
+    assert mslex_quote("a b c") == '"a b c"'
+    assert mslex_quote("C:\\file\\to\\file") == "C:\\file\\to\\file"
```
Comment (harupy) [inline tests/utils/test_string_utils.py:38]: A minimum test to ensure `C:\\file\\to\\file` is not converted to `'C:\\file\\to\\file'`.

**Pull request merged**